base/operators.jl

```diff
@@ -1184,6 +1184,56 @@ end
 
 (f::Fix2)(y) = f.f(y, f.x)
 
+"""
+    Fix{n}(f, x...; kws...)
+    Fix(f; kws...)
+
+A type representing a partially-applied version of a function `f`, with the argument or
+arguments "x" inserted at the `n`th position, and "kws" inserted as keyword arguments.
+In other words, `Fix{3}(f, x1)` behaves similarly to `(y...) -> f(y[1], y[2], x1, y[3:end]...)`.
+
+You may also specify keyword arguments to fix. For example, `Fix(sum; dims=1)`
+would be equivalent to `(a...; k...) -> sum(a...; dims=1, k...)`.
+
+You can also pass multiple arguments at once, which will be inserted after
+the first. You can also fix both arguments and keyword arguments in this way.
+For example, `Fix{2}(g, x1, x2; verbose=true)` would be equivalent to
+`(y...; k...) -> g(y[1], x1, x2, y[2:end]...; verbose=true, k...)`.
+"""
+struct Fix{N,F,T<:Tuple,K<:NamedTuple} <: Function
+    f::F
+    x::T
+    kws::K
+
+    function Fix(f::F; kws...) where {F}
+        xt = ()
+        knt = NamedTuple(kws)
+        new{0,F,_stable_typeof(xt),typeof(knt)}(f, xt, knt)
+    end
+    function Fix{N}(f::F, x, xs...; kws...) where {F,N}
+        xt = (x, xs...)
+        knt = NamedTuple(kws)
+        new{N,F,_stable_typeof(xt),typeof(knt)}(f, xt, knt)
+    end
+    function Fix{N}(f::Type{F}, x, xs...; kws...) where {F,N}
+        xt = (x, xs...)
+        knt = NamedTuple(kws)
+        new{N,F,_stable_typeof(xt),typeof(knt)}(f, xt, knt)
+    end
+end
```

Comment:
Having three internal constructors is ugly.

The constructor specific for `Type` should probably be avoided by relying on `maybeconstructor`/`Constructor`.

In fact, a better design would rely on the autogenerated constructor methods, something like this:

```julia
struct Fix{N<:Tuple{Vararg{Nothing}},F,T<:Tuple,K<:NamedTuple} <: Function
    length_tuple::N
    f::F
    x::T
    kws::K
end
```

This would also provide more type safety, making invalid values of `N` impossible.

Reply:
Not sure I follow, could you give a full example?

Reply:
> The constructor specific for `Type` should probably be avoided by relying on `maybeconstructor`/`Constructor`.

This is borrowed from `Base.Fix1` and `Base.Fix2`